### PR TITLE
Prevent segfaults when using simulatormodule without a simulator

### PR DESCRIPTION
**Pull request by ktbarrett**
A more permanent fix to #1839. When accessing the simulator module without a simulator present, all methods will throw an exception to prevent a possible segfault. `get_sim_precision` is an exception, previous to #1717, this defaulted to `-15`, this behavior has been restored.

@tpoikela

### CODE REVIEW

File: documentation/source/newsfragments/1843.change.rst [hunk 1]
```diff
@@ -0,0 +1 @@
+Simulator Python module now throws an exception if used when no simulator is present. Making the use of cocotb without a simulator present safe.
```
Comment (cmarqu) [inline documentation/source/newsfragments/1843.change.rst:1]:
```suggestion
The :mod:`cocotb.simulator` Python module now throws an exception if used when no simulator is present, making the use of cocotb without a simulator present safe.
```
(Cannot test right now if this will result in the correct link.)

File: cocotb/share/include/cocotb_utils.h [hunk 2]
```diff
@@ -43,6 +43,7 @@ extern void* utils_dyn_open(const char* lib_name);
 extern void* utils_dyn_sym(void *handle, const char* sym_name);
 
 extern int is_python_context;
+extern int simulator_is_active;
```
Comment (eric-wieser) [inline cocotb/share/include/cocotb_utils.h:46]:
```suggestion
extern bool simulator_is_active;
```

File: cocotb/share/lib/embed/gpi_embed.cpp [hunk 3]
```diff
@@ -227,6 +227,8 @@ extern "C" int embed_sim_init(int argc, char const * const * argv)
     if (pEventFn)
         return ret;
 
+    simulator_is_active = 1;
```
Comment (eric-wieser) [inline cocotb/share/lib/embed/gpi_embed.cpp:230]: Can you convince me this is the right place?

Comment (ktbarrett) [reply cocotb/share/lib/embed/gpi_embed.cpp:230]: The line place in the function, or the whole approach? If just the line placement in the function, it's kind of arbitrary, as long as it's set before we load into the Python entry point.

Comment (eric-wieser) [reply cocotb/share/lib/embed/gpi_embed.cpp:230]: The line in the function - ideally this would go right next to the code that initializes `registered_impls`

Comment (ktbarrett) [reply cocotb/share/lib/embed/gpi_embed.cpp:230]: That's not in the function? You mean this should be moved into the GPI, which is a different approach.

The GPI should only ever be loaded by the simulator, so we can assume once we are in the GPI, there is a valid simulator. Because we want to support loading the simulator module outside of the typical flow, the responsibility should rest with the simulator module to prevent itself from using the GPI when it knows it would be invalid.

Moving the flag into the GPI and adding a function to the GPI interface to return that flag means the user must check the function before using any interface. Robust design dictates we shouldn't trust the user and now we have to add error code handling to the entry points of the GPI.

In short, my approach is "No touchy!". I have tested using all the methods and types available on the simulator module from the terminal and ensured they do not segfault. If you think that warrants a test, I'd be happy to add one. I will also add a `simulator.is_running` function. No matter the approach that would be useful.

Comment (eric-wieser) [reply cocotb/share/lib/embed/gpi_embed.cpp:230]: > The GPI should only ever be loaded by the simulator, so we can assume once we are in the GPI, there is a valid simulator.

There are two different bits of the GPI - the C wrapper, which can be loaded by anyone, and the C++ code, which only gets instantiated when the simulator is entered.

I suppose what I'm advocating in the short term is a `gpi_has_registered_impl()` function implemented as `return registered_impls.size() == 0`. We already have `gpi_print_registered_impl()`, so it seems like a reasonable extension

Comment (eric-wieser) [reply cocotb/share/lib/embed/gpi_embed.cpp:230]: Actually, the `gpi_print_registered_impl` function already exposes the information you need via its return value, exposing that in the C extension API would do the trick.

Comment (ktbarrett) [reply cocotb/share/lib/embed/gpi_embed.cpp:230]: > the C wrapper, which can be loaded by anyone

I think that's the crux of the issue. My assumption is that most users know you can't load any library at any time, start using it, and expect it to not blow up; which allows us not do any of the work in the GPI. You seem to suggesting making the GPI interface tolerant against invalid usage in general, which is fine, it's just more work. #1844 seems related to that.

Comment (eric-wieser) [reply cocotb/share/lib/embed/gpi_embed.cpp:230]: I'm suggesting as a short term, we just add a method to the gpi that simply answers the question "will this blow up". We already have one at `gpi_print_registered_impl`, it just prints useless log messages we do not want.

With that method around, you could replace `if (simulator_is_active)` with `if(gpi_has_registered_impl())`. The function call cost (which ultimately just returns a data member of a global anyway) is insignificant compared to the python boundary cost.

Comment (ktbarrett) [reply cocotb/share/lib/embed/gpi_embed.cpp:230]: That's fine. I'll update this PR.

Comment (eric-wieser) [reply cocotb/share/lib/embed/gpi_embed.cpp:230]: Thanks!

File: cocotb/share/lib/simulator/simulatormodule.cpp [hunk 4]
```diff
@@ -744,6 +786,11 @@ static PyObject *get_precision(PyObject *self, PyObject *args)
 {
     COCOTB_UNUSED(self);
     COCOTB_UNUSED(args);
+
+    if (!simulator_is_active) {
+        return PyLong_FromLong(-15);  // preserves old behavior
```
Comment (eric-wieser) [inline cocotb/share/lib/simulator/simulatormodule.cpp:791]: It's probably worth emitting a warning here that no simulator is attached:
```suggestion
        if (PyErr_WarnEx(PyExc_RuntimeWarning, "something about the simulator not actually being attached, assuming whatever precision this is", 1) < 0) {
            return NULL;
        }
        return PyLong_FromLong(-15);  // preserves old behavior
```

File: cocotb/share/include/gpi.h [hunk 5]
```diff
@@ -113,6 +113,13 @@ typedef enum gpi_event_e {
 
 // Functions for controlling/querying the simulation state
 
+/**
+ * Returns 1 if there is a registered GPI implementation.
```
Comment (ktbarrett) [inline cocotb/share/include/gpi.h:117]:
```suggestion
 * Returns 1 if there is a registered GPI implementation, 0 otherwise.
```

File: documentation/source/newsfragments/1843.change.rst [hunk 6]
```diff
@@ -0,0 +1 @@
+The :mod:`cocotb.simulator` Python module now throws an exception if used when no simulator is present, making the use of cocotb without a simulator present safe.
```
Comment (eric-wieser) [inline documentation/source/newsfragments/1843.change.rst:1]:
```suggestion
The :mod:`cocotb.simulator` Python module now throws an :exc`RuntimeError` if used when no simulator is present, making it safe to use cocotb without a simulator present.
```

File: cocotb/share/include/gpi.h [hunk 7]
```diff
@@ -113,6 +113,13 @@ typedef enum gpi_event_e {
 
 // Functions for controlling/querying the simulation state
 
+/**
+ * Returns 1 if there is a registered GPI implementation, 0 otherwise.
+ *
+ * Useful for checking if a simulator is running.
+ */
+int gpi_has_registered_impl(void);
```
Comment (eric-wieser) [inline cocotb/share/include/gpi.h:121]: Any reason not to use bool?
```suggestion
bool gpi_has_registered_impl(void);
```

File: cocotb/share/lib/gpi/GpiCommon.cpp [hunk 8]
```diff
@@ -123,6 +123,11 @@ int gpi_register_impl(GpiImplInterface *func_tbl)
     return 0;
 }
 
+int gpi_has_registered_impl()
```
Comment (eric-wieser) [inline cocotb/share/lib/gpi/GpiCommon.cpp:126]:
```suggestion
bool gpi_has_registered_impl()
```

File: cocotb/share/lib/simulator/simulatormodule.cpp [hunk 9]
```diff
@@ -255,6 +255,12 @@ int handle_gpi_callback(void *user_data)
 static PyObject *log_msg(PyObject *self, PyObject *args)
 {
     COCOTB_UNUSED(self);
+
+    if (!gpi_has_registered_impl()) {
+        PyErr_SetString(PyExc_RuntimeError, "No simulator available!");
+        return NULL;
+    }
+
```
Comment (eric-wieser) [inline cocotb/share/lib/simulator/simulatormodule.cpp:263]:
```suggestion
```
This one is safe without a simulator, right?

File: cocotb/share/lib/simulator/simulatormodule.cpp [hunk 10]
```diff
@@ -810,6 +879,12 @@ static PyObject *deregister(gpi_hdl_Object<gpi_cb_hdl> *self, PyObject *args)
 static PyObject *log_level(PyObject *self, PyObject *args)
 {
     COCOTB_UNUSED(self);
+
+    if (!gpi_has_registered_impl()) {
+        PyErr_SetString(PyExc_RuntimeError, "No simulator available!");
+        return NULL;
+    }
+
```
Comment (eric-wieser) [inline cocotb/share/lib/simulator/simulatormodule.cpp:887]: Again, I don't think `gpilog` needs a simulator.
```suggestion
```

File: cocotb/share/lib/simulator/simulatormodule.h [hunk 11]
```diff
@@ -115,7 +115,12 @@ static PyMethodDef SimulatorMethods[] = {
         "--\n\n"
         "Set the log level for GPI"
     )},
-
+    {"is_running", is_running, METH_NOARGS, PyDoc_STR(
+        "is_running()\n"
+        "--\n\n"
+        "is_running() -> bool\n"
+        "Returns ``True`` if the caller is running within a simulator."
```
Comment (eric-wieser) [inline cocotb/share/lib/simulator/simulatormodule.h:122]: Needs a newsfragment and maybe a `..versionadded`

Comment (ktbarrett) [reply cocotb/share/lib/simulator/simulatormodule.h]: Done. Sorry for the late add. Forgot I wanted this in to address the needs of #1839.

File: documentation/source/newsfragments/1843.feature.rst [hunk 12]
```diff
@@ -0,0 +1 @@
+The :func:`cocotb.simulator.is_running` method was added so a user of cocotb could determine if they are running within a simulator.
```
Comment (eric-wieser) [inline documentation/source/newsfragments/1843.feature.rst:1]:
```suggestion
The :func:`cocotb.simulator.is_running` function was added so a user of cocotb could determine if they are running within a simulator.
```

**Comment by eric-wieser** [comment]
Thanks!
